privacyidea/lib/security/encryptkey.py

```diff
@@ -73,10 +73,14 @@ def hsm_lock(timeout=DEFAULT_TIMEOUT):
         timeout -= 1
         try:
             # Try to get the Lock, since we are acting.
+            log.debug("Requesting lock")
             os.mkdir(DIR)
+            log.debug("Lock acquired")
             yield
             # Cleanup
+            log.debug("Going to release lock")
             os.rmdir(DIR)
+            log.debug("Lock released")
```

Comment:
We should add a `finally` block to delete the lock directory in the end.

Reply:
Giving this a second thought it is not clear to me, if we really need a finally to do this or what we actually try to accomplish with a finally.

I suspect we would only want to ensure, that - if the `yield` raises some exceptions which we do not catch, we want to delete the lock file.
If we go through the loop for 15 seconds and simply can not aquire the lock, then we do not want to release the lock file, since another process very probably still has the lock.

So in my understanding it could look like this:

~~~~python
while timeout > 0:
    timeout =- 1
    try:
        lock()
        i_created_the_lock = True
        yield       # <- the problem is here, if we managed to lock, but bail out.
        release_lock()
        break
    except FileExistsError:
        # We can not get the lock and try again!
        sleep(1)
    except Exception:
         # Catch all other errors that occur in yield?
         if i_created_the_lock:
            release_lock() # Although I have not accomplished anything :-/
~~~~

OK, or we could do it like this:

~~~~python
while timeout > 0:
    timeout =- 1
    try:
        lock()
        i_created_the_lock = True
        yield       # <- the problem is here, if we managed to lock, but bail out.
        break
    except FileExistsError:
        # We can not get the lock and try again!
        sleep(1)
    finally:
         if i_created_the_lock:
            release_lock() # Although I have not accomplished anything :-/
~~~~

Reply:
nicely done in 29a86dc2